### ISSUE TITLE
Constraints not satisfied in complex derivation tree

### ISSUE THREAD

**Issue by andreas-zeller** [rindPHI/isla#36]
I am creating outputs in which a _value_ is preceded by its _length_ (a single byte). In the following grammar, this holds for atoms as well as for sequences:

```python
DUMMY_GRAMMAR = {
    '<start>': ['<sequence>'],
    '<value>': ['<sequence>', '<atom>'],
    '<sequence>': ['S<sequence-length><value>'],
    '<atom>': ['A<atom-length>T'],
    '<sequence-length>': ['<length>'],
    '<atom-length>': ['<length>'],
    '<length>': crange('\x00', '\xff')
}
```

If I add the following constraints, the output produced actually has correct lengths.

```python
DUMMY_CONSTRAINTS = '''
str.to_code(<atom>.<atom-length>) = 1
and
str.to_code(<sequence>.<sequence-length>) = str.len(<sequence>.<value>)
'''
```

```python
solver = ISLaSolver(DUMMY_GRAMMAR, DUMMY_CONSTRAINTS,
                    max_number_free_instantiations=1,
                    max_number_smt_instantiations=1)
tree = solver.solve()
tree
```

Output:

```python
DerivationTree('<start>', (DerivationTree('<sequence>', (DerivationTree('S', (), id=26526), DerivationTree('<sequence-length>', (DerivationTree('<length>', (DerivationTree('\x03', (), id=26682),), id=26683),), id=26684), DerivationTree('<value>', (DerivationTree('<atom>', (DerivationTree('A', (), id=26656), DerivationTree('<atom-length>', (DerivationTree('<length>', (DerivationTree('\x01', (), id=26679),), id=26680),), id=26681), DerivationTree('T', (), id=26658)), id=26611),), id=26528)), id=26436),), id=26435)
```

The `<sequence-length>` field is '\x03', the length of the atom that follows.

If, however, I change the grammar to 

```python
DUMMY_GRAMMAR = {
    '<start>': ['<sequence>'],
    '<value>': ['<sequence>', '<atom>'],
    '<sequence>': ['S<sequence-length><value>'],
    '<atom>': ['A<atom-length><atom-value>'],
    '<atom-value>': ['T', 'F'],  # <-- New
    '<sequence-length>': ['<length>'],
    '<atom-length>': ['<length>'],
    '<length>': crange('\x00', '\xff')
}
```

and leave the constraints unchanged, I get a new (large) tree in which none of the `<sequence-length>` fields matches the length of the successing `<value>`. This is unexpected.

```
DerivationTree('<start>', (DerivationTree('<sequence>', (DerivationTree('S', (), id=27262), DerivationTree('<sequence-length>', (DerivationTree('<length>', (DerivationTree('ÿ', (), id=27427),), id=27428),), ...
```

**Comment by andreas-zeller** [comment]
The same effect also occurs when not using simplified syntax, so it's not a parsing error

```python
DUMMY_CONSTRAINTS = '''
forall <sequence> seq="S{<sequence-length> len}{<value> val}":
    str.to_code(len) = str.len(val)
“'
```

**Comment by rindPHI** [comment]
Could you check whether the problem persists if you add the parameter `enable_optimized_z3_queries=False` to the `ISLaSolver` constructor?

Background: To solve length constraints for "large" numbers (e.g., `str.len(x) = 200`), where Z3 easily times out, ISLa implements a special procedure where first, a number for the length is picked, and second, the solver attempts to create a string of that length and type using grammar expansions, bypassing Z3. However, this can fail, most prominently if a string of that length and type does not exist. You *should* get a RuntimeError with a descriptive message mentioning that parameter, but maybe this fails here.

**Comment by rindPHI** [comment]
This is solved in PR #43. Please still use the setting `enable_optimized_z3_queries=False` mentioned above since the optimization does not work for this complicated use case (you will likely obtain the mentioned RuntimeError.
